pyomo/contrib/solver/gurobi_direct.py

```diff
@@ -0,0 +1,419 @@
+                self._available = GurobiDirect._check_full_license(m)
+            return self._available
+        else:
+            return self.Availability.BadLicense
+
+    @classmethod
+    def _check_full_license(cls, model=None):
+        if model is None:
+            model = gurobipy.Model()
+        model.setParam('OutputFlag', 0)
+        try:
+            model.addVars(range(2001))
+            model.optimize()
+            return cls.Availability.FullLicense
+        except gurobipy.GurobiError:
+            return cls.Availability.LimitedLicense
+
+    def __del__(self):
+        if not python_is_shutting_down():
+            GurobiDirect._num_instances -= 1
+            if GurobiDirect._num_instances == 0:
+                self.release_license()
+
+    @staticmethod
+    def release_license():
+        if gurobipy_available:
+            with capture_output(capture_fd=True):
+                gurobipy.disposeDefaultEnv()
+
+    def version(self):
+        version = (
+            gurobipy.GRB.VERSION_MAJOR,
+            gurobipy.GRB.VERSION_MINOR,
+            gurobipy.GRB.VERSION_TECHNICAL,
+        )
+        return version
+
+    def solve(self, model, **kwds) -> Results:
+        start_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        config = self.config(value=kwds, preserve_implicit=True)
+        if config.timer is None:
+            config.timer = HierarchicalTimer()
+        timer = config.timer
+
+        StaleFlagManager.mark_all_as_stale()
+
+        timer.start('compile_model')
+        repn = LinearStandardFormCompiler().write(
+            model, mixed_form=True, set_sense=None
+        )
+        timer.stop('compile_model')
+
+        if len(repn.objectives) > 1:
+            raise ValueError(
+                f"The {self.__class__.__name__} solver only supports models "
+                f"with zero or one objectives (received {len(repn.objectives)})."
+            )
+
+        timer.start('prepare_matrices')
+        inf = float('inf')
+        ninf = -inf
+        lb = []
+        ub = []
+        for v in repn.columns:
+            _l, _u = v.bounds
+            if _l is None:
+                _l = ninf
+            if _u is None:
+                _u = inf
+            lb.append(_l)
+            ub.append(_u)
+        CON = gurobipy.GRB.CONTINUOUS
+        BIN = gurobipy.GRB.BINARY
+        INT = gurobipy.GRB.INTEGER
+        vtype = [
+            (
+                CON
+                if v.is_continuous()
+                else (BIN if v.is_binary() else INT if v.is_integer() else '?')
+            )
+            for v in repn.columns
+        ]
+        sense_type = '=<>'  # Note: ordering matches 0, 1, -1
+        sense = [sense_type[r[1]] for r in repn.rows]
+        timer.stop('prepare_matrices')
+
+        ostreams = [io.StringIO()] + config.tee
+
+        try:
+            orig_cwd = os.getcwd()
+            if config.working_dir:
+                os.chdir(config.working_dir)
```

Comment:
Do you have to change into that directory for gurobi?

Reply:
Yes.  I haven't found a general "working directory" option for Gurobi.  You can explicitly set the absolute path for the input, log, and results files, but nothing that just generally says "work here".  Changing the CWD is more robust in case Gurobi decided to start writing any additional files.